manubot/cite/csl_item.py

```diff
@@ -1,6 +1,77 @@
 from manubot.cite.citekey import standardize_citekey, infer_citekey_prefix, is_valid_citekey
 
 
+csl_item_type_fixer = {
+    'journal-article': 'article-journal',
+    'book-chapter': 'chapter',
+    'posted-content': 'manuscript',
+    'proceedings-article': 'paper-conference',
+    'standard': 'entry',
+    'reference-entry': 'entry',
+}
+
+def replace_type(key: str) -> str:
+    return csl_item_type_fixer.get(key, key)
+
+
+class CSL_Item(dict):
+    """
+    CSL_Item represents bibliographic information for a single publication.
+
+    On a technical side CSL_Item is a Python dictionary with extra methods
+    that help cleaning and manipulating it.
+
+    These methods relate to:
+    - adding an `id` key and value for CSL item
+    - correcting bibliographic information and its structure   
+    - adding and reading a custom note to CSL item
+
+    """
+    
+    # The ideas for CSL_Item methods come from the following parts of code:
+    #  - [ ] citekey_to_csl_item(citekey, prune=True)
+    #  - [x] csl_item_passthrough
+    #  - [ ] append_to_csl_item_note
+    # The methods in CSL_Item class provide primitives to reconstruct 
+    # fucntions above.
+
+    def __init__(self, dictionary=None, **kwargs):
+        """
+        Can use both a dictionary or keywords to create a CSL_Item object:
+            
+            CSL_Item(title='The Book')
+            CSL_Item({'title': 'The Book'})
+            csl_dict = {'title': 'The Book', 'ISBN': '321-321'}
+            CSL_Item(csl_dict, type='entry')
+            CSL_Item(title='The Book', ISBN='321-321', type='entry')
+            
+        CSL_Item object is usually provided by bibliographic information API, 
+        but constructing small CSL_Item objects is useful for testing.
+        """
+        if dictionary is None:
+            dictionary = dict()            
+        dictionary.update(kwargs)                
```

Comment:
deepcopy (2) is probably the safest, but dictionary (1) would already be clean. To spare creating new auxillary variables we can do this (effectively a 3, but with no in-place effects, hopefully):

```
        if dictionary is None:
            dictionary = dict()            
        super().__init__(dictionary)
        self.update(kwargs)
```

Reply:
I was writing a comment here, but missed posting it - please see a commit. I think it is a cleaner constructor now that does not make inline change for the constructor argument.

Reply:
There is also a small test that shows the constructor argument is left unchanged.

Reply:
The implementation now is a more pure version of option 1. However, I am slightly worried about a situation where objects nested in the input dict get edited in-place, even though the top-level dictionary is new. Therefore, I think deepcopy is probably the way to go?

Reply:
> I am slightly worried about a situation where objects nested in the input dict get edited in-place

Is this a possibility? We should wrap a bibliographic server response into `CSL_Item` at `DOI("blah").csl_item()` and make several transformations over the item or item passing from function to function. Changing the contents of a dict can happen in python, but is this anywhere in the path of transformations?

If you think it is we can have:

```
super().__init__(copy.deepcopy(dictionary)) # safeguards from changes of dictionary contents 
```

but to me it seems overly protective.

Reply:
I've thought about this a bit more and think that we should do either option 2 or 3, but not 1 (which we're currently do). I envision 1 causing problems down the line because the user thinks they've received a new dict but edits a nested object causing problems.

With option 3, we'd clearly document that the underlying dict stays the same. Is it possible to implement it such that the super object becomes the dict that's passed in? Is that bad practice since multiple CSL_Item objects could be sub-instances of the same dict?

If option 3 sounds problematic, I'd say go with option 2 (deepcopy). The proposed implementation looks good to me and ensures none of the arguments passed to the constructor get edited in-place. I think it provides a nice separation... CSL_Item creates a deepcopy/new dict, while all other instance methods edit the existing dict in-place.

> Changing the contents of a dict can happen in python, but is this anywhere in the path of transformations?

I don't know off the top of my head, but this CSL_Item class will be the foundation for many things, possibly used by external applications, so really just thinking of what is best long-term.